### ISSUE TITLE
`network::tests::connections_stress_and_reconnect` fails inconsistently

### ISSUE THREAD

**Issue by hrxi** [nimiq/core-rs-albatross#589]
## New issue checklist

- [x] I have read all of the [`README`](https://github.com/nimiq-network/core/blob/master/README.md) 
- [x] I have searched [existing issues](https://github.com/nimiq-network/core/issues?q=is%3Aissue+sort%3Acreated-desc) and **this is not a duplicate**.

### General information

- Library version(s): master (38a12cd2257a60fb6a96e01ac6552fabfe2b1517)
- Browser version(s): n/a
- Devices/Simulators/Machine affected: n/a
- Reproducible in the testnet? (Yes/No): n/a
- Related issues: none found

#### Expected behavior

All tests pass when I'm running `cargo test`.

#### Actual behavior

```
     Running unittests (target/debug/deps/nimiq_network_libp2p-e99b39114f702d92)

running 14 tests
[...]
test network::tests::connections_stress_and_reconnect ... FAILED
[...]

failures:

---- network::tests::connections_stress_and_reconnect stdout ----
thread 'network::tests::connections_stress_and_reconnect' panicked at 'assertion failed: `(left == right)`
  left: `13`,
 right: `14`', network-libp2p/src/network.rs:1235:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    network::tests::connections_stress_and_reconnect

test result: FAILED. 13 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 15.60s
```

#### Steps to reproduce

In principle, just `cargo test`, but since the failure is not consistent, there's no clear way to do it.

#### Crash log? Screenshots? Videos? Sample project?

See above, I was unfortunately not running with `RUST_BACKTRACE=1`.

**Comment by hrxi** [comment]
Aha, running `stress --cpu 8 --vm 8 --io 8` in the background made the test fail again. (Previously I was compiling lots of stuff in parallel to the test.)

```
running 1 test
test network::tests::connections_stress_and_reconnect ... FAILED

failures:

---- network::tests::connections_stress_and_reconnect stdout ----
thread 'network::tests::connections_stress_and_reconnect' panicked at 'assertion failed: `(left == right)`
  left: `13`,
 right: `14`', network-libp2p/src/network.rs:1235:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/panicking.rs:135:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/panicking.rs:174:5
   4: nimiq_network_libp2p::network::tests::create_network_with_n_peers::{{closure}}
             at ./src/network.rs:1235:13
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/future/mod.rs:84:19
   6: nimiq_network_libp2p::network::tests::connections_stress_and_reconnect::{{closure}}
             at ./src/network.rs:1336:58
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/future/mod.rs:84:19
   8: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/future/future.rs:123:9
   9: tokio::runtime::basic_scheduler::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/basic_scheduler.rs:516:48
  10: tokio::coop::with_budget::{{closure}}
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/coop.rs:102:9
  11: std::thread::local::LocalKey<T>::try_with
             at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/thread/local.rs:413:16
  12: std::thread::local::LocalKey<T>::with
             at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/thread/local.rs:389:9
  13: tokio::coop::with_budget
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/coop.rs:95:5
  14: tokio::coop::budget
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/coop.rs:72:5
  15: tokio::runtime::basic_scheduler::CoreGuard::block_on::{{closure}}::{{closure}}
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/basic_scheduler.rs:516:25
  16: tokio::runtime::basic_scheduler::Context::enter
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/basic_scheduler.rs:374:19
  17: tokio::runtime::basic_scheduler::CoreGuard::block_on::{{closure}}
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/basic_scheduler.rs:515:36
  18: tokio::runtime::basic_scheduler::CoreGuard::enter::{{closure}}
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/basic_scheduler.rs:582:57
  19: tokio::macros::scoped_tls::ScopedKey<T>::set
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/macros/scoped_tls.rs:61:9
  20: tokio::runtime::basic_scheduler::CoreGuard::enter
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/basic_scheduler.rs:582:27
  21: tokio::runtime::basic_scheduler::CoreGuard::block_on
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/basic_scheduler.rs:506:9
  22: tokio::runtime::basic_scheduler::BasicScheduler::block_on
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/basic_scheduler.rs:182:24
  23: tokio::runtime::Runtime::block_on
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/mod.rs:475:46
  24: nimiq_network_libp2p::network::tests::connections_stress_and_reconnect
             at ./src/network.rs:1338:9
  25: nimiq_network_libp2p::network::tests::connections_stress_and_reconnect::{{closure}}
             at ./src/network.rs:1331:11
  26: core::ops::function::FnOnce::call_once
             at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/ops/function.rs:227:5
  27: core::ops::function::FnOnce::call_once
             at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    network::tests::connections_stress_and_reconnect

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 13 filtered out; finished in 14.51s
```

**Comment by jsdanielh** [comment]
Thanks for the issue and the detailed steps on how to reproduce it. It looks like it is a timing issue and I'm proposing the fix in #600 to address it. Using the same steps you posted I didn't see a failure as I used to see it without the change.
